UI/ControlMappingScreen.cpp

```diff
@@ -349,30 +289,148 @@ bool KeyMappingNewKeyDialog::key(const KeyInput &key) {
 	return true;
 }
 
-void KeyMappingNewMouseKeyDialog::CreatePopupContents(UI::ViewGroup *parent) {
+class BindingChoice : public UI::Choice {
+private:
+	std::string bindingText_;
+	bool binding_ = false;
+
+public:
+	BindingChoice(const std::string &text, const std::string& bindingText, UI::LayoutParams *layoutParams = nullptr)
+		: Choice(text, layoutParams), bindingText_(bindingText) {}
+
+	void Draw(UIContext &dc) override;
+	void SetBinding(bool b) { binding_ = b; }
+};
+
+void BindingChoice::Draw(UIContext &dc) {
+	using namespace UI;
+	
+	Style style = dc.theme->itemStyle;
+	if (highlighted_) {
+		style = dc.theme->itemHighlightedStyle;
+	}
+	if (down_) {
+		style = dc.theme->itemDownStyle;
+	}
+	if (HasFocus()) {
+		style = dc.theme->itemFocusedStyle;
+	}
+	if (binding_) {
+		style = dc.theme->itemStyle;
+	}
+	DrawBG(dc, style);
+	
+	if (binding_) {
+		float opacity = sin(5.0*time_now_d())*0.5 + 0.5;
+		uint32_t col = colorAlpha(dc.theme->itemFocusedStyle.background.color, opacity);
+		dc.FillRect(Drawable(col), bounds_);
+	}
+
+	style = dc.theme->itemStyle;
+	dc.SetFontStyle(dc.theme->uiFont);
+
+	const int paddingX = 12;
+	const float availWidth = bounds_.w - paddingX * 2 - textPadding_.horiz();
+	float scale = CalculateTextScale(dc, availWidth);
+
+	dc.SetFontScale(scale, scale);
+
+	Bounds textBounds(bounds_.x + paddingX + textPadding_.left, bounds_.y, availWidth, bounds_.h);
+
+	dc.DrawTextRect(binding_ ? bindingText_.c_str() : text_.c_str(), textBounds, style.fgColor, ALIGN_VCENTER | FLAG_WRAP_TEXT);
+	dc.SetFontScale(1.0f, 1.0f);
+}
+
+void KeyMappingDialog::CreatePopupContents(UI::ViewGroup *parent) {
 	using namespace UI;
 
 	auto km = GetI18NCategory("KeyMapping");
+	auto mc = GetI18NCategory("MappableControls");
+	auto di = GetI18NCategory("Dialog");
+
+	ScrollView *scroll = new ScrollView(ORIENT_VERTICAL, new LinearLayoutParams(FILL_PARENT, 45, 1.0f));
```

Comment:
Specifying a height in layout params for a ScrollView currently doesn't work, but #14779 will fix it.  That said, I don't know if you meant to use 45 - that's pretty short?

-[Unknown]

Reply:
Mhhhh, after #14779 how can I make the popup actually wrap content? It get full screen height even with few item (same goes now for the grid setting in the main screen, but probably less visible giving it have quite some button in it).

Anyway gotta call if for the night, I'll think about it tomorrow :)

Reply:
Fixed the WRAP_CONTENT issue in #14781.  Better for that to be fixed properly so we're not hacking a workaround...

-[Unknown]